### PR TITLE
core/validatorapi: get all validators

### PR DESCRIPTION
**Pull request by dB2510**
Adds support to get all validators from validatorapi if no validator index or pubkey is provided. This issue came up while testing with prysm VC which needs to query beacon node to get the information of total validators active on beacon chain.
Refer: https://github.com/prysmaticlabs/prysm/blob/5c00fcb84ffef6e85d3d105d7288281ae44b44b2/validator/client/wait_for_activation.go#L139

category: misc
ticket: none

### CODE REVIEW

File: core/validatorapi/router.go [hunk 1]
```diff
@@ -1152,10 +1152,6 @@ func getValidatorIDs(query url.Values) []string {
 
 // getValidatorsByID returns the validators with ids being either pubkeys or validator indexes.
 func getValidatorsByID(ctx context.Context, p eth2client.ValidatorsProvider, stateID string, ids ...string) ([]v1Validator, error) {
```
Comment (corverroos) [inline core/validatorapi/router.go:1154]: This will fail in `validatorapi` component in `convertValidators` with "pubshare not found"

**Review by corverroos — CHANGES_REQUESTED** [review]

File: core/validatorapi/router.go (hunk 1 above)
Comment (dB2510) [reply core/validatorapi/router.go:1154]: Thanks for the catch 🙏

So we have 2 options I suppose, either only convert the validators that are there in the cluster with the rest of the validators OR just return ALL validators if no indices/pubkeys are specified with no pubkey<>pubshare conversion.

I would prefer the latter one. What do you think?

Comment (corverroos) [reply core/validatorapi/router.go:1154]: not converting dv public keys will result in the VC not being able to finds it's own validator in the list, resulting in it thinking it is not activated yet. So we have to convert DV public keys to pubshare ALWAYS. 

Suggest somehow detecting that convert should ignore validators if not found in the list. So still erroring by default but allowing only conerting "some" in the case of "querying all validators"

File: core/validatorapi/validatorapi.go [hunk 2]
```diff
@@ -958,7 +958,7 @@ func (c Component) Validators(ctx context.Context, stateID string, validatorIndi
 		return nil, err
 	}
 
-	return c.convertValidators(vals)
+	return c.convertValidators(vals, validatorIndices == nil)
```
Comment (corverroos) [inline core/validatorapi/validatorapi.go:961]: never use nil checks for empty slice checks, always use `len(x) == 0`

File: core/validatorapi/validatorapi.go [hunk 3]
```diff
@@ -989,19 +989,22 @@ func (Component) NodeVersion(context.Context) (string, error) {
 	return fmt.Sprintf("obolnetwork/charon/%s-%s/%s-%s", version.Version, commitSHA, runtime.GOARCH, runtime.GOOS), nil
 }
 
-// convertValidators returns the validator map with all root public keys replaced by public shares.
-func (c Component) convertValidators(vals map[eth2p0.ValidatorIndex]*eth2v1.Validator) (map[eth2p0.ValidatorIndex]*eth2v1.Validator, error) {
+// convertValidators returns the validator map with root public keys replaced by public shares for all validators that are part of the cluster.
+func (c Component) convertValidators(vals map[eth2p0.ValidatorIndex]*eth2v1.Validator, allVals bool) (map[eth2p0.ValidatorIndex]*eth2v1.Validator, error) {
```
Comment (corverroos) [inline core/validatorapi/validatorapi.go:993]: this function should know about its use-cases, rather, variables should be names according to their meaning in the function. So rather `ignoreNotFound bool`

File: testutil/beaconmock/options.go [hunk 4]
```diff
@@ -149,6 +149,12 @@ func WithValidatorSet(set ValidatorSet) Option {
 				}
 			}
 
+			if len(resp) == 0 {
```
Comment (corverroos) [inline testutil/beaconmock/options.go:152]: move this to the top, so first check if "querying for all", in that case return all, else pick from set and log errs.

File: core/validatorapi/validatorapi.go [hunk 5]
```diff
@@ -989,19 +989,22 @@ func (Component) NodeVersion(context.Context) (string, error) {
...
 
-		var ok bool
-		val.Validator.PublicKey, ok = c.getPubShareFunc(val.Validator.PublicKey)
-		if !ok {
+		pubshare, ok := c.getPubShareFunc(val.Validator.PublicKey)
+		if !ok && !ignoreNotFound {
 			return nil, errors.New("pubshare not found")
 		}
+
+		if ok {
```
Comment (corverroos) [inline core/validatorapi/validatorapi.go:1005]: nit: else if (since it is part of the same logic flow)

**Review by corverroos — APPROVED** [review]

**Pull request merged**